activemodel/lib/active_model/attribute_set.rb

```diff
@@ -104,6 +104,10 @@ def reverse_merge!(target_attributes)
     end
 
     def ==(other)
+      if (not defined?(other.class)) || (not self.is_a?(other.class))
+        return false
+      end
+
       attributes == other.attributes
```

Comment:
In case of `AttributeSet.new({})` = `LazyAttributeSet.new({}, {}, {}, {})` we will get NoMethodException, because `attributes` is a protected member of LazyAttributeSet.

Reply:
Yeah, this is unfortunate that `attributes` is a protected method. The changes in this method are too complex: 
`not defined?(other.class)` will always be false because `other.class` is always defined, so this check is not needed.
We need to only check subclasses of the caller, so need `other.is_a?(self.class)`.

So something like:
```ruby
def ==(other)
  other.is_a?(self.class) && 
    attributes == other.send(:attributes) # still that ugly `send`
end
```

Reply:
Ok, fixed

Reply:
Looks like this will change the behavior `lazy_attribute_set == attribute_set` even previously true but now false.

Reply:
Good point! Then `other.is_a?(AttributeSet.class) && attributes == other.send(:attributes)` will satisfy all requirements.